### PR TITLE
feat: conditional coloring for big number chart

### PR DESCRIPTION
**Pull request by gbusch**
### SUMMARY
As suggested here: #21820
allow for conditional formatting (coloring) of Big Number charts.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Screenshot 2023-02-12 at 15 58 36

### TESTING INSTRUCTIONS
* select a "Big Number" chart
* add "conditional formatting" rules in the "Customize" tab

### ADDITIONAL INFORMATION
re-uses existing "ConditionalFormattingControl" from chart table and pivot table

### CODE REVIEW

File: superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts [hunk 1]
```diff
@@ -89,6 +90,45 @@ export default {
...
           },
         ],
+        [
+          {
+            name: 'conditional_formatting',
+            config: {
+              type: 'ConditionalFormattingControl',
+              renderTrigger: true,
+              label: t('Conditional Formatting'),
+              description: t('Apply conditional color formatting to metrics'),
```
Comment (rusackas) [inline superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts:100]: A bit pedantic, but should it just be "metric" in this case?

File: superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts [hunk 2]
```diff
@@ -188,6 +188,22 @@ describe('getColorFunction()', () => {
...
+        targetValueLeft: 50,
+        targetValueRight: 100,
+        colorScheme: '#FF0000',
+        column: 'count',
+      },
+      countValues,
+      false,
+    );
+    expect(colorFunction(50)).toEqual('#FF0000');
+    expect(colorFunction(100)).toEqual('#FF0000');
```
Comment (rusackas) [inline superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts:204]: Would it be worth adding an assertion in the middle of the range (e.g. 75) for good measure?

I'm also not sure if there's another test that asserts against values outside the range (e.g. 25 or 125 in this case). If not, that might be good to add in this test or a new one.

**Comment by rusackas** [comment]
Code looks good, and it seems to work well on the ephemeral environment. The only question I have is whether it's possible/reasonable to make the colors full opacity rather than the sort of pastel colors that they are now. I'm a little worried about contrast/legibility in particular.

**Comment by justin-tomlinson** [comment]
Very useful improvement. It would be great if this was available across the Big Number and Big Number with trend line. Some additional improvements to functionality could be:

- Make the colour scheme be more configurable by using hex codes or RGB values
- Making the colour selection gradient optional. sometimes you want it to be a set colour when it meets the rule rather than be some gradient of the colour. I.e If measure >0.5 then Red, not a gradient of red.

This probably applies to the conditional formatting on the table chart too. -

**Comment by rusackas** [comment]
@gbusch curious about your input on any of the above comments... would love to see this PR get through! :)

File: superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts (hunk 1 above)
Comment (gbusch) [reply superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts]: true, done

File: superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts (hunk 2 above)
Comment (gbusch) [reply superset-frontend/packages/superset-ui-chart-controls/test/utils/getColorFormatters.test.ts:206]: added tests, please resolve it ok

File: superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts [hunk 3]
```diff
@@ -89,6 +90,45 @@ export default {
...
+              },
+              mapStateToProps(explore, _, chart) {
+                const verboseMap = explore?.datasource?.hasOwnProperty(
+                  'verbose_map',
+                )
+                  ? (explore?.datasource as Dataset)?.verbose_map
+                  : explore?.datasource?.columns ?? {};
+                const { colnames, coltypes } =
+                  chart?.queriesResponse?.[0] ?? {};
+                const numericColumns =
```
[the remark this answers is not in the document]
Comment (gbusch) [reply superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberTotal/controlPanel.ts:112]: might make sense but would prefer not to do it here

**Comment by gbusch** [comment]
Sorry, was busy and then almost forgot about it...

Colors are defined in this file: 
superset-frontend/src/explore/components/controls/ConditionalFormattingControl/FormattingPopoverContent.tsx

I renamed the colors to make it clearer that they refer to the theme colors and added possibility for darker colors (from theme). Does this make sense?

**Comment by gbusch** [comment]
@rusackas What do you think?
Pipeline is red... Don't understand why, as the changes are rather small.

**Comment by kgabryje** [comment]
> Pipeline is red... Don't understand why, as the changes are rather small.

The E2E tests are sometimes flaky... I restarted the CI, hopefully it passes this time

**Review by rusackas — APPROVED** [review]

**Comment by mtrentz** [comment]
Big Numbers doesn't get color formatted when its value is zero. Anyone else experience this?

If I set a rule for "error when < 5" and the value is 0, nothing happens.

If the value is 0.0001 then it colors.

**Comment by rusackas** [comment]
> Big Numbers doesn't get color formatted when its value is zero. Anyone else experience this?
> 
> If I set a rule for "error when < 5" and the value is 0, nothing happens.
> 
> If the value is 0.0001 then it colors.

Would you be willing to open a PR or an Issue about this? We haven't heard much outcry, and would love any help you can provide to set it on a course toward resolution.

**Comment by mtrentz** [comment]
@rusackas opened this one #28038

**Pull request merged**